### ISSUE TITLE
Synapse creates its network but does not connect to it on first launch

### ISSUE THREAD

**Issue by cdavernas** [serverlessworkflow/synapse#15]
**What happened**:

Synapse creates its network but does not connect to it on first launch, resulting in workers not being able to contact it using its advertized host name (i.e. synapse)

**What you expected to happen**:

Synapse connects to its network after creating it, in case it does not exist

**How to reproduce it**:

1. Delete the synapse network if it already exist
2. Start Synapse
3. Deploy and start a workflow
4. The deployed worker crashes upon startup, failling to resolve the synapse DNS

**Environment**:

Docker

**Workaround**:

1. Start Synapse, wait for it to be ready and accepting requests
2. Restart Synapse
3. Voila!

**Comment by bvandewe** [comment]
Still cant get this to work on Mac OS 10.14.6... 

```
~$ docker image ls -a | grep synapse
ghcr.io/serverlessworkflow/synapse-worker                          0.1.0      87889178eadb   4 hours ago     277MB
ghcr.io/serverlessworkflow/synapse-worker                          latest     87889178eadb   4 hours ago     277MB
ghcr.io/serverlessworkflow/synapse                                 latest     2407099d2971   4 hours ago     376MB
~$
~$ docker rm synapse
synapse
~$ docker network rm synapse
synapse
~$ docker run -d --name synapse -e "SYNAPSE_API_HOSTNAME=synapse" -v /var/run/docker.sock:/var/run/docker.sock --add-host=host.docker.internal:host-gateway -p 42286:42286 -p 41387:41387 ghcr.io/serverlessworkflow/synapse:latest
966cb0ec438c6ee6d14557eac0e718bedd22ae7fbf71f35cd9869d7a4980eae5
~$
~$ docker ps -a | grep synapse
966cb0ec438c   ghcr.io/serverlessworkflow/synapse:latest                                "dotnet Synapse.Serv…"   15 seconds ago   Up 14 seconds                80/tcp, 0.0.0.0:41387->41387/tcp, 8080/tcp, 0.0.0.0:42286->42286/tcp    synapse
~$
~$ docker restart synapse
synapse
~$
~$ docker ps -a | grep synapse
966cb0ec438c   ghcr.io/serverlessworkflow/synapse:latest                                "dotnet Synapse.Serv…"   32 seconds ago   Exited (139) 1 second ago                                                                            synapse
~$
~$ docker logs synapse
[04/15/2022 12:44:01] info: ProtoBuf.Grpc.Server.ServicesExtensions.CodeFirstServiceMethodProvider[0]
      RPC services being provided by Synapse.Apis.Management.Grpc.SynapseGrpcManagementApi: 16
[04/15/2022 12:44:01] info: ProtoBuf.Grpc.Server.ServicesExtensions.CodeFirstServiceMethodProvider[0]
      RPC services being provided by Synapse.Apis.Runtime.Grpc.SynapseGrpcRuntimeApi: 20
[04/15/2022 12:44:01] warn: Microsoft.AspNetCore.Server.Kestrel[0]
      Overriding address(es) 'http://+:80'. Binding to endpoints defined via IConfiguration and/or UseKestrel() instead.
[04/15/2022 12:44:01] info: Microsoft.Hosting.Lifetime[14]
      Now listening on: http://[::]:41387
[04/15/2022 12:44:01] info: Microsoft.Hosting.Lifetime[14]
      Now listening on: http://[::]:42286
[04/15/2022 12:44:01] info: Microsoft.Hosting.Lifetime[0]
      Application started. Press Ctrl+C to shut down.
[04/15/2022 12:44:01] info: Microsoft.Hosting.Lifetime[0]
      Hosting environment: Production
[04/15/2022 12:44:01] info: Microsoft.Hosting.Lifetime[0]
      Content root path: /app/
[04/15/2022 12:44:27] info: Microsoft.Hosting.Lifetime[0]
      Application is shutting down...
[04/15/2022 12:44:29] info: ProtoBuf.Grpc.Server.ServicesExtensions.CodeFirstServiceMethodProvider[0]
      RPC services being provided by Synapse.Apis.Management.Grpc.SynapseGrpcManagementApi: 16
[04/15/2022 12:44:29] info: ProtoBuf.Grpc.Server.ServicesExtensions.CodeFirstServiceMethodProvider[0]
      RPC services being provided by Synapse.Apis.Runtime.Grpc.SynapseGrpcRuntimeApi: 20
Unhandled exception. System.ArgumentNullException: Value cannot be null. (Parameter 'parameters')
   at Docker.DotNet.NetworkOperations.Docker.DotNet.INetworkOperations.ConnectNetworkAsync(String id, NetworkConnectParameters parameters, CancellationToken cancellationToken)
   at Synapse.Runtime.Services.DockerRuntimeHost.ExecuteAsync(CancellationToken stoppingToken) in /src/src/runtime/Synapse.Runtime.Docker/Services/DockerRuntimeHost.cs:line 77
   at Microsoft.Extensions.Hosting.Internal.Host.StartAsync(CancellationToken cancellationToken)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.RunAsync(IHost host, CancellationToken token)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.RunAsync(IHost host, CancellationToken token)
   at Program.<Main>$(String[] args) in /src/src/apps/Synapse.Server/Program.cs:line 82
   at Program.<Main>(String[] args)
~$
```

On Mac OS Mojave:

```
~$ sw_vers
ProductName:	Mac OS X
ProductVersion:	10.14.6
BuildVersion:	18G9323
~$ 
~$ docker info
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc., v0.7.1)
  compose: Docker Compose (Docker Inc., v2.2.1)
  scan: Docker Scan (Docker Inc., v0.11.0)

Server:
 Containers: 9
  Running: 1
  Paused: 0
  Stopped: 8
 Images: 17
 Server Version: 20.10.11
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 7b11cfaabd73bb80907dd23182b9347b4245eb5d
 runc version: v1.0.2-0-g52b36a2
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
  cgroupns
 Kernel Version: 5.10.76-linuxkit
 Operating System: Docker Desktop
 OSType: linux
 Architecture: x86_64
 CPUs: 10
 Total Memory: 3.841GiB
 Name: docker-desktop
 ID: SAME:3O77:4JIU:HONE:YRKQ:TFB6:JMQ2:WXK4:JBTA:6AXN:HDPO:5HTW
 Docker Root Dir: /var/lib/docker
 Debug Mode: true
  File Descriptors: 53
  Goroutines: 60
  System Time: 2022-04-15T12:50:09.7457618Z
  EventsListeners: 3
 HTTP Proxy: http.docker.internal:3128
 HTTPS Proxy: http.docker.internal:3128
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: true
 Insecure Registries:
 Live Restore Enabled: false
```

Will test on Linux and revert... 

Thanks, this is super exciting!! :)

**Comment by cdavernas** [comment]
@bvandewe I'll investigate. This is wierd, there's no way the `parameters` arg can be null, as it can be seen in sources.
Plus, Im not facing the problem in both Win 11 x64 and WSL 2.

**Comment by bvandewe** [comment]
Tried the latest image, same result...

```

 ~$ docker image inspect ghcr.io/serverlessworkflow/synapse | grep synapse@sha256
            "ghcr.io/serverlessworkflow/synapse@sha256:8dfd4ab6c4cc2f013beb3fdabf4cc764889797ef2c4d0486e13556ebc17a5ac1"
 ~$
 ~$ docker network rm synapse
synapse
 ~$ docker rm synapse
synapse
 ~$ sudo docker run -d --name synapse -e "SYNAPSE_API_HOSTNAME=synapse" -v /var/run/docker.sock:/var/run/docker.sock --add-host=host.docker.internal:host-gateway -p 42286:42286 -p 41387:41387 ghcr.io/serverlessworkflow/synapse:latest
Password:
659e786beb124b8d9a4d38e54e2438f8a0a65330e938bed853e279ba2dc6222d
 ~$
 ~$ docker ps -a | grep synapse
659e786beb12   ghcr.io/serverlessworkflow/synapse:latest                                "dotnet Synapse.Serv…"   3 seconds ago   Up 2 seconds                 80/tcp, 0.0.0.0:41387->41387/tcp, 8080/tcp, 0.0.0.0:42286->42286/tcp    synapse
 ~$
 ~$ sudo docker restart synapse
synapse
 ~$ docker ps -a | grep synapse
659e786beb12   ghcr.io/serverlessworkflow/synapse:latest                                "dotnet Synapse.Serv…"   17 seconds ago   Exited (139) Less than a second ago                                                                           synapse
 ~$ 
 ~$ docker logs synapse
[04/15/2022 13:15:41] info: ProtoBuf.Grpc.Server.ServicesExtensions.CodeFirstServiceMethodProvider[0]
      RPC services being provided by Synapse.Apis.Management.Grpc.SynapseGrpcManagementApi: 16
[04/15/2022 13:15:41] info: ProtoBuf.Grpc.Server.ServicesExtensions.CodeFirstServiceMethodProvider[0]
      RPC services being provided by Synapse.Apis.Runtime.Grpc.SynapseGrpcRuntimeApi: 20
[04/15/2022 13:15:41] warn: Microsoft.AspNetCore.Server.Kestrel[0]
      Overriding address(es) 'http://+:80'. Binding to endpoints defined via IConfiguration and/or UseKestrel() instead.
[04/15/2022 13:15:41] info: Microsoft.Hosting.Lifetime[14]
      Now listening on: http://[::]:41387
[04/15/2022 13:15:41] info: Microsoft.Hosting.Lifetime[14]
      Now listening on: http://[::]:42286
[04/15/2022 13:15:41] info: Microsoft.Hosting.Lifetime[0]
      Application started. Press Ctrl+C to shut down.
[04/15/2022 13:15:41] info: Microsoft.Hosting.Lifetime[0]
      Hosting environment: Production
[04/15/2022 13:15:41] info: Microsoft.Hosting.Lifetime[0]
      Content root path: /app/
[04/15/2022 13:15:54] info: Microsoft.Hosting.Lifetime[0]
      Application is shutting down...
[04/15/2022 13:15:56] info: ProtoBuf.Grpc.Server.ServicesExtensions.CodeFirstServiceMethodProvider[0]
      RPC services being provided by Synapse.Apis.Management.Grpc.SynapseGrpcManagementApi: 16
[04/15/2022 13:15:56] info: ProtoBuf.Grpc.Server.ServicesExtensions.CodeFirstServiceMethodProvider[0]
      RPC services being provided by Synapse.Apis.Runtime.Grpc.SynapseGrpcRuntimeApi: 20
Unhandled exception. System.ArgumentNullException: Value cannot be null. (Parameter 'parameters')
   at Docker.DotNet.NetworkOperations.Docker.DotNet.INetworkOperations.ConnectNetworkAsync(String id, NetworkConnectParameters parameters, CancellationToken cancellationToken)
   at Synapse.Runtime.Services.DockerRuntimeHost.ExecuteAsync(CancellationToken stoppingToken) in /src/src/runtime/Synapse.Runtime.Docker/Services/DockerRuntimeHost.cs:line 77
   at Microsoft.Extensions.Hosting.Internal.Host.StartAsync(CancellationToken cancellationToken)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.RunAsync(IHost host, CancellationToken token)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.RunAsync(IHost host, CancellationToken token)
   at Program.<Main>$(String[] args) in /src/src/apps/Synapse.Server/Program.cs:line 82
   at Program.<Main>(String[] args)
```

**Comment by cdavernas** [comment]
@bvandewe I cannot reproduce. Asked colleagues to try running a clean install on Docker, it worked perfectly 😲

**Comment by bvandewe** [comment]
Weird, i get the same `System.ArgumentNullException: Value cannot be null. (Parameter 'parameters')` on Ubuntu 18.04 & 20.04 as well...

**Comment by cdavernas** [comment]
@bvandewe can you confirm that, as discussed in PMs, the bug is due to conflicting port assignment on your system?
Note that with next release, the server will no longer expose ports 80 and 8080 (but will instead expose ports 42286 and 41387)

**Comment by tsurdilo** [comment]
Not able to reproduce this issue on
MacOS Big Sur 11.2.3
Docker version 20.10.8, build 3967b7d
Docker Desktop 4.0.1

Will try on my M1 as well

**Comment by cdavernas** [comment]
Looking at the official .NET Docker client library used by Synapse, I see that the thrown exception is misleading: it says that `parameters` arg is null, though it seems it's the `Container` property which is. I opened https://github.com/dotnet/Docker.DotNet/issues/559 to address that.
Knowing that, it seems that the container id returned by the `TryGetDockerContainerId` is null on some systems, even though it's occurring on Docker, with the exact same image that works on others. 
It is probably be related to the following command, but I still need to get my hand on a system where I can repro the issue:
```
bash -c "head -1 /proc/self/cgroup|cut -d/ -f3"
```

**Comment by cdavernas** [comment]
Should be fixed by 96d5d119cac2faae3a8ee76c833f5b38050ea25f
